File: p784m/letter_case_permutation.py
from typing import List


class Solution:
    def letterCasePermutationV1(self, S: str) -> List[str]:
        ans = []

        def backtrack(path: str, i: int) -> None:
            nonlocal S
            while i < len(S) and S[i].isdigit():
                path += S[i]
                i += 1
            if i == len(S):
                ans.append(path)
                return
            backtrack(path + S[i].lower(), i + 1)
            backtrack(path + S[i].upper(), i + 1)

        backtrack("", 0)
        return ans

    def letterCasePermutationV2(self, S: str) -> List[str]:
        ans = [""]
        for ch in S:
            if ch.isalpha():
                ans = [prefix + c for prefix in ans for c in [ch.lower(), ch.upper()]]
            else:
                ans = [prefix + ch for prefix in ans]
        return ans


# TESTS
for S, expected in [
    ("a1b2", ["a1b2", "a1B2", "A1b2", "A1B2"]),
    ("3z4", ["3z4", "3Z4"]),
    ("12345", ["12345"]),
    ("0", ["0"]),
]:
    sol = Solution()
    actual = sol.letterCasePermutationV1(S)
    print("Letter case permutation of", S, "->", actual)
    assert actual == expected
    assert expected == sol.letterCasePermutationV2(S)

File: p784m/letter_case_permutation.js
/**
 * @param {string} S
 * @return {string[]}
 */
var letterCasePermutation = function (S) {
  let ans = [];
  // Closure S and ans
  let recBacktrack = function (prefix, s) {
    if (s.length === 0) {
      ans.push(prefix);
      return;
    }
    if (s[0] >= '0' && s[0] <= '9') {
      recBacktrack(prefix + s[0], s.substr(1));
    } else {
      recBacktrack(prefix + s[0].toLowerCase(), s.substr(1));
      recBacktrack(prefix + s[0].toUpperCase(), s.substr(1));
    }
  };
  recBacktrack('', S);
  return ans;
};
// TEST
[
  ['a1b2', ['a1b2', 'a1B2', 'A1b2', 'A1B2']],
  ['3z4', ['3z4', '3Z4']],
  ['12345', ['12345']],
  ['0', ['0']],
].forEach(([S, expected]) => {
  const actual = letterCasePermutation(S);
  console.log('Letter case permutation of', S, '->', actual);
  console.assert(actual.toString() === expected.toString());
});
